polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1716,6 +1717,268 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
+
+  LogicalResult matchAndRewrite(sycl::SYCLHostHandlerSetKernel op,
+                                PatternRewriter &rewriter) const final {
+    // `set_kernel` op can only be raised in a CGF.
+    auto enclosingFunc = op->getParentOfType<LLVM::LLVMFuncOp>();
+    assert(enclosingFunc);
+    auto handler = op.getHandler();
+    assert(handler == getHandler(enclosingFunc));
+
+    // Discover `set_nd_range` ops in the CGF. We are currently only interested
+    // in `parallel_for` launches, hence there should be exactly one
+    // `set_nd_range` op in the function.
+    auto setNDRangeOps =
+        enclosingFunc.getOps<sycl::SYCLHostHandlerSetNDRange>();
+    if (!llvm::hasNItems(setNDRangeOps, 1))
+      return failure();
+    auto range = *setNDRangeOps.begin();
+
+    // Extract the number of parameters and the pointer to the
+    // `kernel_signatures` data structure from annotations in the current block.
+    auto failureOrParamInfo = getKernelParameterInfo(op->getBlock());
+    if (failed(failureOrParamInfo))
+      return failure();
+    auto [numParams, paramDesc] = *failureOrParamInfo;
+
+    // Parse `kernel_signatures` to get the parameter kinds.
+    auto failureOfParamKinds = getParameterKinds(numParams, paramDesc);
+    if (failed(failureOfParamKinds))
+      return failure();
+    auto paramKinds = *failureOfParamKinds;
+
+    // Adjust `num_params` if the last entry is a sentinel.
+    if (numParams > 0 && paramKinds.back() == kernel_param_kind_t::kind_invalid)
+      --numParams;
+
+    // No point in trying to raise a `schedule_kernel` without arguments.
+    if (numParams == 0)
+      return failure();
+
+    // Discover `set_captured` ops in the CGF. We check that we have captured a
+    // value matching the expected kind for each parameter.
+    auto setCapturedOps = enclosingFunc.getOps<sycl::SYCLHostSetCaptured>();
+    unsigned numCaptured = llvm::range_size(setCapturedOps);
+
+    // Wrong number of `set_captured` ops.
+    if (numCaptured != numParams)
+      return failure();
+
+    SmallVector<Value> args(numCaptured);
+    SmallVector<Type> syclTypes(numCaptured, rewriter.getNoneType());
+    for (auto capture : setCapturedOps) {
+      unsigned idx = capture.getIndex();
+      // Index is out of bounds.
+      if (idx >= numCaptured)
+        return failure();
+
+      args[idx] = capture.getValue();
+
+      auto syclType = capture.getSyclType();
+      auto kind = paramKinds[idx];
+      switch (kind) {
+      case kernel_param_kind_t::kind_accessor:
+        // We expected an accessor, but the captured value isn't one.
+        if (!syclType.has_value() || !isa<sycl::AccessorType>(syclType.value()))
+          return failure();
+
+        syclTypes[idx] = syclType.value();
+        break;
+
+      case kernel_param_kind_t::kind_std_layout:
+        // We expected standard layout, but the captured value is a
+        // SYCL-specific type.
```

Comment:
Do all SYCL types have a special `kernel_param_kind_t`? I could imagine `sycl::vec` to be passed as a standard layout argument for example.

Reply:
Ah yes, I changed the check to look specifically for accessors passed allegedly passed as standard layout params.